assembly/src/tokens/stream.rs

```diff
@@ -25,24 +26,44 @@ impl<'a> TokenStream<'a> {
             return Err(ParsingError::empty_source());
         }
         let mut tokens = Vec::new();
-        let mut doc_comments = BTreeMap::new();
+        let mut proc_comments = BTreeMap::new();
+        let mut module_comment = None;
 
-        let mut comment = Comment(None);
+        let mut comment_builder = CommentBuilder(None);
 
         for line in source.lines() {
             let line = line.trim();
             if line.starts_with(DOC_COMMENT_PREFIX) {
-                comment.append_line(line);
-            } else if line.is_empty() && !comment.is_empty() {
-                return Err(ParsingError::malformed_doc_comment(tokens.len()));
-            } else {
-                if !comment.is_empty() {
-                    doc_comments.insert(tokens.len(), comment.take_content());
+                comment_builder.append_line(line);
+            } else if line.starts_with(LINE_COMMENT_PREFIX) {
+                continue;
+            } else if line.is_empty() {
+                if !comment_builder.is_empty() {
+                    if tokens.is_empty() && module_comment.is_none() {
+                        // if we haven't read any tokens yet, but already have built a comment, a
+                        // new line must indicate the end of a module comment.
+                        module_comment = comment_builder.take_comment();
+                    } else {
+                        // if procedure comment is not immediately followed by anything.
```

Comment:
nit: I would change the comment here to: "since we already have a module comment, this is a procedure comment which is followed by a blank line."